system/security/plugins/jwtSecurity/jwtSecurity.cpp

```diff
@@ -0,0 +1,1129 @@
+     *
+     * @return  New string with all characters uppercased.
+     */
+    static std::string str_toupper(std::string s)
+    {
+        std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c){return std::toupper(c);});
+
+        return s;
+    }
+
+    /**
+     * Validates an encoded JWT token (as if presented as a byte stream from an
+     * external service).  Tokens are validated according to the OpenID Connect
+     * specification (https://openid.net/specs/openid-connect-core-1_0.html#IDToken).
+     * All REQUIRED items are, indeed, required.
+     *
+     * If token is validated then a copy of the token and all claims
+     * are cached under the username of the current user.
+     *
+     * @param   token           Encoded token
+     * @param   subject         Username of current user; will be compared
+     *                          to the 'sub' value within the token; case-
+     *                          sensitive
+     * @param   refreshToken    Refresh token; will stuffed into the user
+     *                          cache if token verifies without errors
+     * @param   nonce           The random string supplied to the JWT
+     *                          endpoint that presumably returned the
+     *                          encoded token; will be compared to the
+     *                          'nonce' value within the token; case-
+     *                          sensitive; this may be an empty string;
+     *                          if an empty string, no nonce value will
+     *                          be verified
+     *
+     * @return  true if all validation checks pass, false otherwise
+     */
+    bool verifyToken(const std::string& token, const std::string& subject, const std::string& refreshToken, const std::string& nonce, const std::string& endpointURL)
+    {
+        bool    verificationResult = false;
+
+        try
+        {
+            CollectedScopePerms     wuScopePerms;
+            CollectedScopePerms     fileScopePerms;
+
+            // Decode the token string
+            auto decodedToken = jwt::decode(token);
+
+            // Create the JWT token verifier with known claims
+            auto jwtVerifier = jwt::verify()
+                                .with_subject(subject)
+                                .with_audience(clientID.str());
+
+            // Add an issuer claim only if an endpoint value was provided
+            if (!endpointURL.empty())
+                jwtVerifier = jwtVerifier.with_issuer(endpointURL);
+
+            // Add a nonce claim only if a nonce value was provided
+            if (!nonce.empty())
+                jwtVerifier = jwtVerifier.with_claim("nonce", jwt::claim(nonce));
+
+            // Append signature algorithm verifier
+            std::string included_algo = str_toupper(decodedToken.get_algorithm());
+            if (included_algo == "HS256")
+                jwtVerifier = jwtVerifier.allow_algorithm(jwt::algorithm::hs256(keyContents));
```

Comment:
I've heard of attacks where you are configured with a public key but the token says it uses a symmetric algorithm... which it used the public key to encrypt.. i.e. the mismatch allowed it to fool the receiver that it was a valid token.  Do we know if that will work here?  If not perhaps we can recognize if we are using a symmetric or public key and restrict the algorithms?

Reply:
That is clever.  I've added a check for public key format, then limiting the algorithms checked appropriately.  Great idea!